gufe/storage/externalresource.py

```diff
@@ -0,0 +1,113 @@
+import hashlib
+import warnings
+import pathlib
+
+from typing import ClassVar
+
+
+class ExternalResourceError(Exception):
+    """Base class for errors due to problems with external resources"""
+    # TODO: is it necessary to have a base class here? Would you ever have
+    # one catch that handles both subclass errors?
+
+
+class MissingExternalResourceError(ExternalResourceError):
+    """Error when the external resource could not be loaded"""
+
+
+class ChangedExternalResourceError(ExternalResourceError):
+    """Error when there's a SHA256 mismatch with an external resource"""
+
+
+class ExternalStorage:
+    """Abstract base for external storage."""
+
+    allow_changed: ClassVar[bool] = False
+
+    def validate(self, location, sha2):
+        if not self.get_sha2(location) == sha2:
+            msg = (f"Hash mismatch for {location}: this object "
+                   "may have changed.")
+            if not self.allow_changed:
+                # NOTE: having it here instead of in a DataLoader means that
+                # you can only change it for the whole system, instead of
+                # for a specific object
+                raise ChangedExternalResourceError(
+                    msg + " To allow this, set ExternalStorage."
+                    "allow_changed = True"
+                )
+            else:
+                warnings.warn(msg)
+
+    def get_sha2(self, location: str):
+        filelike = self._as_filelike(location, bytes_mode=True)
+        sha2 = hashlib.sha256()
```

Comment:
So sha256 is better, but md5 is faster.  We're not really worried about collisions or security, so can we pick speed here?